quinn-proto/src/connection/assembler.rs

```diff
@@ -164,16 +175,26 @@ impl Assembler {
 
     /// Offset after the largest byte received
     pub(crate) fn limit(&self) -> u64 {
-        self.recvd.max().map_or(0, |x| x + 1)
+        self.limit
     }
 
-    /// Whether we've received all data below `final_offset`
-    ///
-    /// `final_offset` must be greater than the offset of any received data.
-    pub(crate) fn is_complete_at(&self, final_offset: u64) -> bool {
-        debug_assert!(final_offset >= self.limit());
-        (self.recvd.len() == 1 && self.recvd.iter().next().unwrap() == (0..final_offset))
-            || final_offset == 0
+    /// Whether all data prior to `self.limit()` has been read
+    pub(crate) fn is_fully_read(&self) -> bool {
+        if let State::Unordered { ref recvd } = self.state {
```

Comment:
Hmm, maybe write this as a `match` expression instead? Would improve the symmetry and obviate the fall-through case (which I think is not very obvious here).

Reply:
Yeah, that's definitely clearer.